File: leetcode/Flatten_Binary_Tree_to_Linked_List.java
 /**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public void flatten(TreeNode root) {
        if (root == null) return;
        Stack<TreeNode> s = new Stack<TreeNode>();
        TreeNode cn = root;
        while (true) {
            if (cn.left == null && cn.right == null) {
                if (!s.isEmpty())
                    cn = cn.right = s.pop();
                else
                    break;
            } else if (cn.left != null) {
                if (cn.right != null) s.push(cn.right);
                cn.right = cn.left;
                cn.left = null;
                cn = cn.right;
            } else {
                cn = cn.right;
            }
        }
    }
}

File: leetcode/Flatten_Binary_Tree_to_Linked_List.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode *last;
    void doFlatten(TreeNode *node) {
        if (node == NULL)
            return;
        last = node;
        doFlatten(node->left);
        if (node->left != NULL) {
            last->right = node->right;
            node->right = node->left;
            node->left = NULL;  // forgot to reset this at first
        }
        doFlatten(last->right);
    }
public:
    void flatten(TreeNode *root) {
        doFlatten(root);
    }
};
